### PR TITLE
Add support for chip.ble.commisioning for CHIP-REPL

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem
Python scripting API does not yet support ble commisioning

 #### Summary of Changes
This builds on #5054 and adds support for BLE wifi commisioning (thread not yet added but should be easy to add).

This example places my ESP32 on wifi and prints out 'Device paired':

```py
import chip.ble.commissioning

c = chip.ble.commissioning.Connect(discriminator=3840, pin=12345678)
if c.needsNetworkCredentials:
    c.ConnectToWifi('myssid', 'mypassword')

if not c.paired:
    print('Device is NOT paired')
else:
    print('Device paired')
```

### CODE REVIEW

**Review by saurabhst — APPROVED** [review]

**Comment by bzbarsky-apple** [comment]
@andy31415 Could you please resolve the merge conflicts?  That would also hopefully remove all the extraneous changes from the PR, making it much simpler to review...

**Comment by andy31415** [comment]
@bzbarsky-apple - merged.

File: src/controller/python/chip/internal/commisioner.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+#
+from chip.configuration import GetLocalNodeId
+from chip.native import NativeLibraryHandleMethodArguments, GetLibraryHandle
+from ctypes import c_uint64, c_uint32, c_uint16, c_char_p
+from typing import Optional
+from chip.internal.types import NetworkCredentialsRequested, OperationalCredentialsRequested, PairingComplete
+import ctypes
+
+# Not using c_void_p directly is IMPORTANT. Python auto-casts c_void_p
+# to intergers and this can cause 32/64 bit issues.
```
Comment (mspang) [inline src/controller/python/chip/internal/commisioner.py:24]: typo: integers

Also WTF ctypes :(

Good find.

File: src/controller/python/chip/internal/commisioner.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+def OnOperationalCredentialsRequested(csr, csr_length):
+    GetCommisioner()._OnOperationalCredentialsRequested(ctypes.string_at(csr, csr_length))
+
+@PairingComplete
+def OnPairingComplete(err: int):
+    GetCommisioner()._OnPairingComplete(err)
+
+
+# Commisioner states regarding pairing
+COMMISIONER_INITIALIZED = 0
```
Comment (mspang) [inline src/controller/python/chip/internal/commisioner.py:42]: Use python enum?

File: src/controller/python/chip/internal/CommisionerImpl.cpp [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+class ServerStorageDelegate : public chip::PersistentStorageDelegate
+{
+public:
+    void SetDelegate(chip::PersistentStorageResultDelegate * delegate) override { mAsyncDelegate = delegate; }
+
+    void GetKeyValue(const char * key) override
+    {
+        uint8_t buffer[kMaxKeyValueSize];
+        uint16_t bufferSize = sizeof(buffer) + 1;
```
Comment (bzbarsky-apple) [inline src/controller/python/chip/internal/CommisionerImpl.cpp:36]: Why `+ 1`?  This does not seem right.  Is that supposed to be  `- 1`?

File: src/controller/python/chip/internal/CommisionerImpl.cpp [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+extern "C" chip::Controller::DeviceCommissioner * pychip_internal_Commisioner_New(uint64_t localDeviceId)
+{
+    std::unique_ptr<chip::Controller::DeviceCommissioner> result;
+    CHIP_ERROR err;
+
+    chip::python::ChipMainThreadScheduleAndWait([&]() {
+        result = std::make_unique<chip::Controller::DeviceCommissioner>();
+
+        // System and Inet layers exlicitly passed to indicate that the CHIP stack is
```
Comment (bzbarsky-apple) [inline src/controller/python/chip/internal/CommisionerImpl.cpp:198]:
```suggestion
        // System and Inet layers explicitly passed to indicate that the CHIP stack is
```

File: src/controller/python/chip/internal/ChipThreadWork.h [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+#include <core/CHIPError.h>
+
+namespace chip {
+namespace python {
+
+using WorkCallback = std::function<void()>;
+
+/// Schedules a task to be run on the CHIP main thread and waits for that
+/// task to be complete
```
Comment (bzbarsky-apple) [inline src/controller/python/chip/internal/ChipThreadWork.h:30]: Maybe document explicitly that `WorkCallback` is guaranteed to NOT run after `ChipMainThreadScheduleAndWait` returns?

**Review by bzbarsky-apple — APPROVED** [review]

File: src/controller/python/chip/internal/commisioner.py (hunk 1 above)
Comment (andy31415) [reply src/controller/python/chip/internal/commisioner.py:24]: This was painful to debug.

**Comment by andy31415** [comment]
Ready for another review pass. I double-checked that I could place my ESP32 on wifi via the python script.

TIL to spell commissioning correctly ...

File: src/controller/python/chip/internal/CommisionerImpl.cpp [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+    void GetKeyValue(const char * key) override
+    {
+        uint8_t buffer[kMaxKeyValueSize];
+        uint16_t bufferSize = sizeof(buffer) + 1;
+        CHIP_ERROR err      = GetKeyValue(key, buffer, bufferSize);
+
+        if (err == CHIP_NO_ERROR)
+        {
+            buffer[bufferSize] = 0;
+            mAsyncDelegate->OnValue(key, reinterpret_cast<const char *>(buffer));
```
[the remark this answers is not in the document]
Comment (andy31415) [reply src/controller/python/chip/internal/CommisionerImpl.cpp]: I was forced to provide an implementation of a storage delegate, however the example I based this on had only the sync version (and I believe we have other places where only the async version works).

I am unclear on what the final goal is - I believe we should have a central API and python should be be required to have a custom implementation.

Added a TODO on this.
Also do not like the naming - it is not obvious that `SetValue(const char *, const char *)` is async but `SetValue(const char *, const void *, size_t)` is sync - I would rather expect both to be the same and the void* to allow for binary data.

**Review by woody-apple — APPROVED** [review]

**Comment by woody-apple** [comment]
@jelderton ?
@andy31415 can you take a peek at the merge conflicts?

**Review by mspang — APPROVED** [review]

**Pull request merged**